node-wasm/src/ports.rs

```diff
@@ -0,0 +1,394 @@
+        let MultiplexMessage::<Command> { id, mut payload } =
+            from_value(ev.data()).context("could not deserialize message")?;
+
+        if let Some(port) = ev.get_ports().into_iter().take(1).last() {
+            if let Command::Management(WorkerCommand::ConnectPort(maybe_port)) = &mut payload {
+                let _ = maybe_port.insert(port);
+            }
+        };
+
+        Ok(MultiplexMessage { id, payload })
+    }
+}
+
+impl TryFrom<MessageEvent> for MultiplexMessage<WorkerResult> {
+    type Error = Error;
+
+    fn try_from(ev: MessageEvent) -> Result<Self, Self::Error> {
+        let MultiplexMessage::<WorkerResult> { id, mut payload } =
+            from_value(ev.data()).context("could not deserialize message")?;
+
+        if let Some(port) = ev.get_ports().into_iter().take(1).last() {
+            if let Ok(WorkerResponse::Subscribed(maybe_port)) = &mut payload {
+                let _ = maybe_port.insert(port);
+            }
+        };
+
+        Ok(MultiplexMessage { id, payload })
+    }
+}
+
+// Instead of supporting communication with just `MessagePort`, allow using any object which
+// provides compatible interface, eg. `Worker`
+#[wasm_bindgen]
+extern "C" {
+    /// Abstraction over JavaScript MessagePort (but also runtime.Port for browser extension).
+    /// Object which can `postMessage` and receive `onmessage` events.
+    #[derive(Debug)]
+    pub(crate) type MessagePortLike;
+
+    #[wasm_bindgen(catch, method, structural, js_name = postMessage)]
+    fn post_message(this: &MessagePortLike, message: &JsValue) -> Result<(), JsValue>;
+
+    #[wasm_bindgen(catch, method, structural, js_name = postMessage)]
+    pub fn post_message_with_transferable(
+        this: &MessagePortLike,
+        message: &JsValue,
+        transferable: &JsValue,
+    ) -> Result<(), JsValue>;
+}
+
+impl From<MessagePort> for MessagePortLike {
+    fn from(value: MessagePort) -> Self {
+        MessagePortLike { obj: value.into() }
+    }
+}
+
+pub(crate) fn split_port<Tx>(port: MessagePortLike) -> Result<(PortSender<Tx>, RawPortReceier)> {
+    let _post_message: Function = Reflect::get(&port, &"postMessage".into())?
+        .dyn_into()
+        .context("could not get object's postMessage")?;
+
+    let port = Rc::new(port);
+
+    let (tx, receiving_channel) = mpsc::unbounded_channel();
+    let onmessage = Closure::new(move |ev: MessageEvent| {
+        if tx.send(ev).is_err() {
+            error!("forwarding message from port failed, no receiver waiting");
+        }
+    });
+    register_onmessage_callback(port.as_ref(), &onmessage)?;
+
+    let sender = PortSender::<Tx> {
+        port: port.clone(),
+        _sent: PhantomData,
+    };
+    let receiver = RawPortReceier {
+        port,
+        receiving_channel,
+        onmessage,
+    };
+    Ok((sender, receiver))
+}
+
+// Sending part of a wrapped MessagePort.
+// WARN: do not use unit structs as a sent type, as they are interpreted as a
```

Comment:
as a what? :p

Reply:
I stumbled upon a fix for handling unit structs as I was writing this comment, so I've abandoned it and then didn't go back to remove it :see_no_evil: Removed, as unit structs are not handled correctly